### ISSUE TITLE
EMQX memory leak (RAM usage increases over time)

### ISSUE THREAD

**Issue by RodaBVB** [emqx/emqx#3492]
**Environment**:
- EMQ X version:  Tests performed with **emqx V4.0.2, emqx V4.0.7** and **emqx V4.1-rc2**
- Hardware configuration:  aws c5.9xl (36vCPUs/72GB RAM)
- OS: AWS Ubuntu 16.04 server
- Erlang/OTP version: **Erlang/OTP 18**

**Topology**:
- AWS network load balancer (NLB)
- 4x EMQX nodes in cluster (36 vCPUs, 72GB RAM each) Instance type: c5.9xl

**Configuration**

```
log.level = critical
clustering = static
cluster.static.seeds = emqx@192.168.x.x, emqx@192.168.x.x, emqx@192.168.x.x, emqx@192.168.x.x
acl_cache_tll = 60m
mqtt.max_packet_size = 100MB
```
Publishers connecting to external listeners (8 ports in range 1883-1890):
```
listener.tcp.external = 0.0.0.0:1883
listener.tcp.external.acceptors = 64
listener.tcp.external.max_connections = 1024000
listener.tcp.external.max_conn_rate = 1
listener.tcp.external.active_n = 100
listener.tcp.external.zone = external
listener.tcp.external.access.1 = allow all
listener.tcp.external.backlog = 1024
listener.tcp.external.send_timeout = 15s
listener.tcp.external.send_timeout_close = on
listener.tcp.external.nodelay = true
listener.tcp.external.reuseaddr = true
```
Consumers connecting on internal listener (port 11883):
```
listener.tcp.internal = 0.0.0.0:11883
listener.tcp.internal.acceptors = 64
listener.tcp.internal.max_connections = 1024000
listener.tcp.internal.max_conn_rate = 1000
listener.tcp.internal.active_n = 1000
listener.tcp.internal.zone = internal
listener.tcp.internal.backlog = 512
listener.tcp.internal.send_timeout = 5s
listener.tcp.internal.send_timeout_close = on
listener.tcp.internal.recbuf = 64KB
listener.tcp.internal.sndbuf = 64KB
listener.tcp.internal.nodelay = false
listener.tcp.internal.reuseaddr = true
```
Zones config:
```
zone.external.idle_timeout = 15s
zone.external.enable_acl = on
zone.external.enable_ban = on
zone.external.enable_stats = on
zone.external.acl_deny_action = ignore
zone.external.force_gc_policy = 16000|16MB
zone.external.keepalive_backoff = 0.75
zone.external.max_subscriptions = 0
zone.external.upgrade_qos = off
zone.external.max_inflight = 32
zone.external.retry_interval = 30s
zone.external.max_awaiting_rel = 100
zone.external.await_rel_timeout = 300s
zone.external.session_expiry_interval = 2h

zone.external.max_mqueue_len = 10000
zone.external.mqueue_priorities = none
zone.external.mqueue_default_priority = highest
zone.external.mqueue_store_qos0 = false
zone.external.enable_flapping_detect = off
zone.external.use_username_as_clientid = false
zone.external.ignore_loop_deliver = false
zone.external.strict_mode = false

zone.internal.allow_anonymous = true
zone.internal.enable_stats = on
zone.internal.enable_acl = off
zone.internal.acl_deny_action = ignore
zone.internal.max_subscriptions = 0
zone.internal.max_inflight = 128
zone.internal.max_awaiting_rel = 1000
zone.internal.max_mqueue_len = 10000
zone.internal.mqueue_store_qos0 = true
zone.internal.enable_flapping_detect = off
zone.internal.ignore_loop_deliver = false
zone.internal.strict_mode = false
zone.internal.bypass_auth_plugins = true
```
Full configuration file: [emqx.conf](https://github.com/emqx/emqx/files/4680631/emqx.conf.txt)

Plugins used:
```
emqx_management.
emqx_rule_engine.
emqx_recon.
emqx_retainer.
emqx_dashboard.
emqx_auth_http.
emqx_auth_redis.
```
## Test
**Test run**
- Connecting 800k clients (publishers), publishing at rate of ~20k msgs/second
- Connecting 2k clients (consumers) each subscribing to wildcard topics for handling publishers messages.
- Test duration:  1 days, 16 hours, 34 minutes

## What happened and what you expected to happen
* **RAM memory dropped gradually over time for each of 4 emqx nodes in cluster:**
  * RAM drop avg 50.97 - 48.73 ~= 2,24
  * RAM drop min 47.88 - 46.54v ~= 1,34
  * RAM drop max 54,82 - 53,16 ~= 1,66

![emqxRAM40hrs](https://user-images.githubusercontent.com/20912951/82872421-52d41280-9f33-11ea-862d-822bef48c1fc.jpeg)

![Screenshot from 2020-05-26 10-19-12](https://user-images.githubusercontent.com/20912951/82877398-69ca3300-9f3a-11ea-8dc3-87b758119c93.png)

**- CPU usage was at ~35% with spikes to max 49%:**

![cpu](https://user-images.githubusercontent.com/20912951/82872727-b6f6d680-9f33-11ea-883c-83430ef85b69.jpeg)

**- Connection count on emqx node:**
![conn](https://user-images.githubusercontent.com/20912951/82872827-dd1c7680-9f33-11ea-9a55-df30e71c01d5.jpeg)

**Anything else we need to know?**:

- SSL is handled by the AWS NLB, and connections proxied to EMQX nodes as insecure #3289 
- No messages were enqueued or dropped for any of the clients https://github.com/emqx/emqx/issues/3451#issuecomment-626115503
- mqueue_store_qos0 = false for publisher clients

All of the available information point to issues with enqueued messages, or slow subscribers. In our case all messages are being acknowledged in time, no messages drop or enqueues are happening but still, a constant trend of increasing RAM memory usage is observed. 
Why is this happening? 

Thanks, Stefan.

**Comment by HJianBo** [comment]
Hi, @RodaBVB  Thank you for so detailed testing.

Can you run it again and get some diagnostic data for me?

1. The completely `emqx.conf`
2. Execute the following commands to get memory diagnostic data. On the `Beginning of workload started`, `Middle of test time`, `Before ending tests`:
    * Cmds for VM memory usage details
    ```
    ./bin/emqx_ctl vm memory
    ./bin/emqx_ctl recon memory
    ./bin/emqx_ctl recon allocated
    ```

    * Execute the `bin/emqx attach` to enter the erlang console, then execute the followings to get VM fragmentation info:
    ``` erlang
    rp(recon_alloc:fragmentation(current)).
    ```

In addition, you can execute the `./bin/emqx_ctl recon bin_leak` to Force Garbage Collection  before ending tests. Observe the memory difference before and after the command executed and send the command output to me

**Comment by HJianBo** [comment]
Any news?

**Comment by RodaBVB** [comment]
Sorry for late answer.

Currently we are not able to use above mentioned environment setup. 
Instead the following local environment is used:

**Environment:**

* EMQ X version: **emqx 4.0.7**
* Hardware configuration: 8vCPUs/16GB RAM
* OS: 16.04.3 LTS (Xenial Xerus)
* Kernel: Linux 4.15.0-101-generic
* Erlang/OTP version: Erlang/OTP 22 [erts-10.6.2]

**Topology:**

* 7x physical machines for running 324k clients in total.
* 1x NGINX 1.16 (least connection routing strategy)
* 2x EMQX nodes in cluster (162k clients per one EMQX node)

**Issue is reproduced.**

## Test
**Test run**
- 324k clients (publishers), publishing at rate of ~8k msgs/second
- 42 clients (consumers) each subscribing to wildcard topics for handling publishers messages.
- Test duration:  1 days, 20 hours

## What happened and what you expected to happen

* **RAM memory dropped gradually over time for each of 2 emqx nodes in cluster:**
  * RAM drop avg 5.55GB - 5.21GB ~= 0.34GB

![Broker1_1](https://user-images.githubusercontent.com/20912951/83627933-007c9c80-a598-11ea-9596-c60806b28d5c.png)
![Broker1_2](https://user-images.githubusercontent.com/20912951/83627944-06727d80-a598-11ea-88fd-d5dc35655d1c.png)

**- CPU usage was at ~32% with spikes to max 55%:**

![Broker1_cpu](https://user-images.githubusercontent.com/20912951/83628525-eee7c480-a598-11ea-87f1-905fda292bf5.png)

**- Connection count on emqx node:**

![Broker1_conenctions](https://user-images.githubusercontent.com/20912951/83628520-ec856a80-a598-11ea-9bf8-b7d9e305d1ef.png)


**- All clients got connected on 1st June at 18h 20mins.**

During test run, the following commands were executed 3 times at:
Day 1 - 1st Jun at 18h 25min
Day 2 - 2nd Jun at 14h 50min
Day 3 - 3rd Jun at 10h 15min



./bin/emqx_ctl vm memory

| Parameter             | Day 1      | Day 2      | Day 3      |
|-----------------------|------------|------------|------------|
| memory/total          | 6998422808 | 6678691208 | 6739190440 |
| memory/processes      | 2943531448 | 2619655528 | 2672932680 |
| memory/processes_used | 2943362328 | 2619561576 | 2672766432 |
| memory/system         | 4054891360 | 4059035680 | 4066257760 |
| memory/atom           | 1689801    | 2353481    | 2566513    |
| memory/atom_used      | 1663712    | 2324729    | 2547266    |
| memory/binary         | 316095360  | 314565000  | 321212824  |
| memory/code           | 25955335   | 25955335   | 25955335   |
| memory/ets            | 3294544976 | 3298255432 | 3298244784 |

./bin/emqx_ctl recon memory

| Parameter         | Day 1             | Day 2             | Day 3             |
|-------------------|-------------------|-------------------|-------------------|
| usage/current     | 0.81611944658214  | 0.786002044619602 | 0.781514194621586 |
| usage/max         | 0.851747814377897 | 0.836497548011424 | 0.83081248921575  |
| used/current      | 6991510568        | 6828420224        | 6852242848        |
| used/max          | 9034720872        | 9322921872        | 9412369496        |
| allocated/current | 8568430592        | 8688001024        | 8767057920        |
| allocated/max     | 10607271936       | 11145187328       | 11329114112       |
| unused/current    | 1575568888        | 1858259824        | 1914949936        |
| unused/max        | 1572551064        | 1822265456        | 1916744616        |

./bin/emqx_ctl recon allocated


| Parameter            | Day 1      | Day 2      | Day 3      |
|----------------------|------------|------------|------------|
| binary_alloc/current | 491814912  | 387743744  | 497319936  |
| driver_alloc/current | 253788160  | 251691008  | 251691008  |
| eheap_alloc/current  | 3508604928 | 3525685248 | 3471867904 |
| ets_alloc/current    | 3780935680 | 3993796608 | 4002185216 |
| fix_alloc/current    | 278953984  | 283934720  | 290226176  |
| ll_alloc/current     | 131334144  | 131334144  | 131334144  |
| sl_alloc/current     | 294912     | 294912     | 294912     |
| std_alloc/current    | 118521856  | 120356864  | 120356864  |
| temp_alloc/current   | 1179648    | 1179648    | 1179648    |
| binary_alloc/max     | 1085882368 | 1160855552 | 1170817024 |
| driver_alloc/max     | 261390336  | 261390336  | 262438912  |
| eheap_alloc/max      | 4794916864 | 5002379264 | 5135290368 |
| ets_alloc/max        | 3828383744 | 4058546176 | 4090003456 |
| fix_alloc/max        | 303857664  | 315392000  | 317489152  |
| ll_alloc/max         | 131334144  | 131334144  | 131334144  |
| sl_alloc/max         | 46039040   | 59674624   | 66125824   |
| std_alloc/max        | 142901248  | 142901248  | 142901248  |
| temp_alloc/max       | 12566528   | 12713984   | 12713984   |

In attached file all of three command outputs from both brokers can be found.

[Additional outputs](https://github.com/emqx/emqx/files/4722786/metrics.zip)

>1. The completely emqx.conf

The full emqx.conf can be found in the first post.

>2. Execute the bin/emqx attach to enter the erlang console, then execute the followings to get VM fragmentation info:

Since attaching to erlang console causes emqx to crash if sagnificant number of mqtt clients are connected, we have not extracted VM fragmentation info.

Also, manually running the GC is not done. We do have recon plugin enabled that should run GC periodically. 

Please elaborate on the data provided, as we can see similar behavior between AWS and local test.
RAM usage is still gradually increasing.

**Comment by RodaBVB** [comment]
Update:
Broker status after 7 days of test:
![Screenshot from 2020-06-08 10_47_09](https://user-images.githubusercontent.com/20912951/84027702-b0367d80-a98f-11ea-9208-651c5103697a.png)

The trend of RAM usage is obvious.

**Comment by RodaBVB** [comment]
Any updates on this issue?

**Comment by Kracker1911** [comment]
@RodaBVB Don't subscribe to wildcard topics. Clients subscribing to the full topics, will make emqx broker performance better.

**Comment by ipapp** [comment]
> @RodaBVB Don't subscribe to wildcard topics. Clients subscribing to the full topics, will make emqx broker performance better.

@Kracker1911, your comment doesn't seem to refer to the topic of this ticket - memory leak. The information you shared is correct, it was addressed by other tickets. Still, I don't see the connection with the memory leak problem.
 
EMQX team, do you have any feedback on this issue?

**Comment by RodaBVB** [comment]
Hello @HJianBo , any update?

Issue still present: **EMQX 4.0.7** 

We are still testing emqx with 800k clients, the memory usage gradually increases over time. 

I would appreciate it if you would give me some feedback on this issue.

See graphics below: 

![Screenshot from 2020-07-23 16-43-28](https://user-images.githubusercontent.com/20912951/88300468-cca23580-cd03-11ea-8836-c7397c3f9658.png)
![Screenshot from 2020-07-23 16-43-42](https://user-images.githubusercontent.com/20912951/88300476-ce6bf900-cd03-11ea-8a73-d2642ecb6ae8.png)

**Comment by HJianBo** [comment]
Hi, @RodaBVB Sorry for the late reply.

First, there are some questions.

> 324k clients (publishers), publishing at the rate of ~8k msgs/second
42 clients (consumers) each subscribing to wildcard topics for handling publishers messages.

0. The QoS level for the `~8k msgs/s`?
1. Will these consumers be able to process these PUBLISH messages in a timely manner? That looks like about 1k msgs/sec? How did you make sure the message wasn't queued up in Broker?

    These messages maybe piled up into the Subscriber process's memory, if the Subscriber processes are busy.  You can get it by `./bin/emqx_ctl recon proc_count message_queue_len 10`, This command will show the 10 processes with the highest Erlang process message queue length.
    
    And you can force shutdown it to stop the unrestrained growth of this memory. see `zone.external.force_shutdown_policy` at emqx.conf

**Comment by terry-xiaoyu** [comment]
@RodaBVB I think the memory was not `leaked`. It might be occupied by the erlang memory allocators due to memory fragmentation in allocators, as the result shown by `emqx_ctl vm memory` was not keeping increasing. 

I don't think this is a problem unless the memory is continuing increase and then the system crashes. The Erlang VM will reclaim the memory at some point according to the current allocation strategy used. See https://erlang.org/doc/man/erts_alloc.html#strategy

Change the allocation strategy to `address order best fit` will reduce the memory fragmentation, but it may or may not  improve the problem in your case, as there's still enough free memory. You could try to add the following line to etc/vm.args to use the `address order best fit` strategy:

```
+Muasaobf
```

And then restart the emqx.

**Comment by RodaBVB** [comment]
In order to verify the assumption that EMQX will release the occupied memory, we did an additional test. This time we occupied RAM on instances where EMQX is located. The steps were the following:
1. Connect all the clients (~720K)
2. Using the "stress" Linux tool, occupy memory leaving around 1GB free
3. Monitor RAM usage over 10 days

The outcome is the following:
![Screenshot from 2020-08-17 11-17-39](https://user-images.githubusercontent.com/7778135/90385845-f9880500-e083-11ea-8d53-25c3c3d3290b.png)
![Screenshot from 2020-08-17 11-17-52](https://user-images.githubusercontent.com/7778135/90385853-fbea5f00-e083-11ea-86bb-502bd1676d76.png)

Showing stable operation for over 10 days.
Thanks for the support.
